### ISSUE TITLE
Version 3.1 Array pop_front doesn't return a value

### ISSUE THREAD

**Issue by PelegMedia** [godotengine/godot#23144]
**Godot version:**
v3.1.alpha.calinou

**OS/device including version:**
Windows 7

**Issue description:**
I'm trying to use a shuffled array to get random list of coordination pairs.
I'm sending a function an array, which contains a list of [x,y] items.
When calling pop_front the method should return the first item in the array.
The editor shows an error that there is no return type and also the method popup shows its return type is void

**Minimal reproduction project:**
It's simple.
Just create an array with few dummy items and call pop_front.

Version 3.1 and pop_front is defined with return value void.

The first one is a parameter to a function:
arrListOfCoordinations : Array
The second one is a local array:
var arrPos : Array = []

![image](https://user-images.githubusercontent.com/39774781/47226613-b26a5380-d3c9-11e8-883f-4a6ae4fd4f10.png)
![image](https://user-images.githubusercontent.com/39774781/47226633-b9916180-d3c9-11e8-8340-ba6019039188.png)

**Comment by PelegMedia** [comment]
Hi @pgruenbacher 

I did more tests and it fails in two cases:
1. The var getting the result has a hint in its declaration.
2. You try to cast the result with a converting function such as int()


The following example works:
```
func _ready():
	var arr1 : Array 
	
	arr1.append(1)
	arr1.append(2)
	arr1.append(3)
	
	print (arr1.pop_front())
```

The following examples fail.

arr2 has array hint
```
func _ready():
	var arr1 : Array 
	var arr2 : Array
	
	arr1.append([1,2])
	arr1.append([1,2])
	arr1.append([1,2])
	
	arr2 = arr1.pop_front()
	print (arr2)
```
Giving me:
![image](https://user-images.githubusercontent.com/39774781/47233051-33314b80-d3da-11e8-97f4-f5ccd7b3684b.png)

arr2 has an int hint:
```
func _ready():
	var arr1 : Array 
	var arr2 : int
	
	arr1.append(1)
	arr1.append(1)
	arr1.append(2)
	
	arr2 = int(arr1.pop_front())
	print (arr2)
```
Giving:
![image](https://user-images.githubusercontent.com/39774781/47233173-8e633e00-d3da-11e8-9beb-262221b895b8.png)

result is castes into an int:
```
func _ready():
	var arr1 : Array 
	var arr2 
	
	arr1.append(1)
	arr1.append(1)
	arr1.append(2)
	
	arr2 = int(arr1.pop_front())
	print (arr2)
```
Giving:
![image](https://user-images.githubusercontent.com/39774781/47233340-0cbfe000-d3db-11e8-8ea4-02dfe92247c2.png)

If I remove the hint and the cast function it will work OK.
This is the major issue.

The second one is minor. The return value in this command's tool-tip should be "variant" instead of "void".

**Comment by pgruenbacher** [comment]
ah. i see the issue ok thank you. 
https://github.com/godotengine/godot/blob/b17e71b6e5e035f49b5b3b5b55b9cdac80215d72/core/variant_call.cpp#L1702

This isn't a bug per se. I don't think the variant functions that can return a variant aren't properly being declared here... hard to say... but I do think this is impacting pop_back as well. will have to experiment later.

**Comment by pgruenbacher** [comment]
This isn't a typo bug though, this is just a limitation of the typehints at the moment for any function that can return a variant (Dictionary, Array, etc.)

**Comment by pgruenbacher** [comment]
and if we were to implement dynamic type checking on these variant returns, that would lead to major speed decrease in the gdscript, so at that point, there's no real benefit for the type hint in the first place.

**Comment by pgruenbacher** [comment]
@PelegMedia I'm pretty sure this is unfixable until typehints allow for typed array declarations like this : "Array[]" which declares an Array of ARrays. LIkewise Object[] would be Array of objects". But that's not implemented yet. Regardless we should probably document this as part of the gdscript typehint docs page as a thing to watch out for.

**Comment by akien-mga** [comment]
Duplicate of #22791?

**Comment by akien-mga** [comment]
Assuming duplicate of #22791, fixed by #22899. Please comment if it's not the case.
